### PR TITLE
Parse HTML List

### PR DESCRIPTION
**Pull request by sixdouglas**
Fixing the #288 
Trying to reproduce the bug

### CODE REVIEW

**Comment by sixdouglas** [comment]
Screenshot 2019-10-29 at 20 58 13
Here is the result in the generated PDF

**Comment by sixdouglas** [comment]
Not reproducing the problem so far

**Comment by rammetzger** [comment]
In the following example, HTMLWorker.parseToList() should return a list of items, with some items containing a list of items.
We can not use HtmlParser.parse() because we need to edit the result list before writing to PDF document.
```
final String htmlText = 
    "<html>"
  + "<body>"
  + "<p>What should you say?</p>"
  + "<ul>"
  + "  <li>Hello</li>"
  + "  <li>World</li>"
  + "</ul>"
  + "<ol>"
  + "  <li>Element-1"
  + "    <ol>"
  + "      <li>Element-1-1</li>"
  + "      <li>Element-1-2</li>"
  + "    </ol>"
  + "  </li>"
  + "  <li>Element-2"
  + "    <ol>"
  + "      <li>Element-2-1</li>"
  + "      <li>Element-2-2</li>"
  + "    </ol>"
  + "  </li>"
  + "</ol>"
  + "</body>"
  + "</html>";
		
final StringReader reader = new StringReader(htmlText);

final StyleSheet styleSheet = new StyleSheet();
// styleSheet.loadTagStyle(...)

final Map<String, Object> interfaceProps = new HashMap<>();
// interfaceProps.put(...);

final List<com.lowagie.text.Element> elements = HTMLWorker.parseToList(reader, styleSheet, interfaceProps);

```
The expected result is:
```
+---------------------------+----------------------------+
| What should you say       | type()==Element.LISTITEM   |
+---------------------------+----------------------------+
| Hello                     | type()==Element.LISTITEM   |
+---------------------------+----------------------------+
| World                     | type()==Element.LISTITEM   |
+---------------------------+----------------------------+
| Element-1                 | type()==Element.LIST       |
|     +---------------------+--------------------------+ +
|     | Element-1-1         | type()==Element.LISTITEM | |
|     +---------------------+--------------------------+ +
|     | Element-1-2         | type()==Element.LISTITEM | |
|     +---------------------+--------------------------+ +
+---------------------------+----------------------------+
| Element-2                 | type()==Element.LIST       |
|     +---------------------+--------------------------+ +
|     | Element-2-1         | type()==Element.LISTITEM | |
|     +---------------------+--------------------------+ +
|     | Element-2-2         | type()==Element.LISTITEM | |
|     +---------------------+--------------------------+ +
+---------------------------+----------------------------+
```
But the current result is:
```
+---------------------------+----------------------------+
| What should you say       | type()==Element.LISTITEM   |
+---------------------------+----------------------------+
| Hello                     | type()==Element.LISTITEM   |
+---------------------------+----------------------------+
| World                     | type()==Element.LISTITEM   |
+---------------------------+----------------------------+
| Element-1                 | type()==Element.LISTITEM   |
+---------------------------+----------------------------+
| Element-2                 | type()==Element.LISTITEM   |
+---------------------------+----------------------------+
```

**Comment by sixdouglas** [comment]
You're right. With your piece of code the nested list is correctly identified.
I've committed it in this PR though I think you should do it under your own name 😉. If so please close this PR and submit a new one with all the samples. It helps improve the project.
Otherwise I think you can approve the PR.

**Comment by rammetzger** [comment]
Thank you for accepting my suggestion for improvement. I have no idea what the processes are to make changes to projects at Github. I would have to learn all this first. That's why I think it's better if you make those changes for me. I hope we do not find any more bugs. OpenPDF is an excellent tool.

File: pdf-toolbox/src/test/java/com/lowagie/examples/html/ParseHelloHtml.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+     */
+    public static void main(String[] args) {
+        System.out.println("Parse Hello World");
+
+        // step 1: creation of a document-object
+        try (Document document = new Document()) {
+            PdfWriter.getInstance(document, new FileOutputStream("parseHelloWorld.pdf"));
+            // step 3: we open the document
+            document.open();
+            // step 2:
```
Comment (noavarice) [inline pdf-toolbox/src/test/java/com/lowagie/examples/html/ParseHelloHtml.java]: Step 1: "description"
Step **3**: "description"
Step **2**: "no description"
...
(⊙_☉)

File: pdf-toolbox/src/test/java/com/lowagie/examples/html/ParseNestedHtmlList.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+    /**
+     * Generates an HTML page with the text 'Hello World'
+     *
+     * @param args no arguments needed here
+     */
+    public static void main(String[] args) {
+        System.out.println("Parse Nested HTML List");
+        try {
+            final String htmlText =
```
Comment (noavarice) [inline pdf-toolbox/src/test/java/com/lowagie/examples/html/ParseNestedHtmlList.java]: What's the difference between test above and this one?

File: README.md [hunk 3]
```diff
@@ -101,9 +101,9 @@ Significant [Contributors to OpenPDF](https://github.com/LibrePDF/OpenPDF/graphs
   [@lapo-luchini](https://github.com/lapo-luchini)  
   [@jeffrey-easyesi](https://github.com/jeffrey-easyesi)  
   [@V-F](https://github.com/V-F)     
-  [@sixdouglas](https://github.com/sixdouglas) - Douglas Six
+  [@sixdouglas](https://github.com/sixdouglas) - Douglas Six
```
Comment (noavarice) [inline README.md]: Spaces?

Comment (sixdouglas) [reply README.md]: The spaces here are meant to be displayed as new line in the Markdown file. There are some at the end of every other lines for this purpose.

File: pdf-toolbox/src/test/java/com/lowagie/examples/html/ParseNestedHtmlList.java (hunk 2 above)
Comment (sixdouglas) [reply pdf-toolbox/src/test/java/com/lowagie/examples/html/ParseNestedHtmlList.java]: This one uses `HTMLWorker#parseToList()` with a StringReader and produce a list on items.
The previous one uses `HtmlParser#parse()` with a Document to produce a PDF.
 
The first one produced a good PDF document with the nested HTML List. But the problem of this issue is really about the `HTMLWorker#parseToList()` not producing the nested List as expected.

Event if the first test, producing the PDF, is not showing the bug, I choose to keep it as a Non-Regression Test.

Comment (noavarice) [reply pdf-toolbox/src/test/java/com/lowagie/examples/html/ParseNestedHtmlList.java]: But why second example contains HTML markup as a String and not as a file?

**Comment by rammetzger** [comment]
I checked it with 1.3.13-SNAPSHOT and it works fine. Nested lists are displayed as expected.

But there is another issue with lost lines (https://github.com/LibrePDF/OpenPDF/issues/317) that isn’t solved yet.


Von: Andreas Rosdal <notifications@github.com>
Gesendet: Donnerstag, 16. Januar 2020 07:33
An: LibrePDF/OpenPDF <OpenPDF@noreply.github.com>
Cc: Michael Metzger | RA Consulting <M.Metzger@rac.de>; Comment <comment@noreply.github.com>
Betreff: Re: [LibrePDF/OpenPDF] Parse HTML List (#290)


Is this ready now?

**Pull request merged**